### ISSUE TITLE
Flutter engine crashes on some Android devices due to "Failed to setup Skia Gr context"

### ISSUE THREAD

**Issue by yrom** [flutter/flutter#47804]
## Steps to Reproduce

This crash was happened on users' Android phone. I don't know clear steps to reproduce.

User starts the `FlutterActivity` then the APP crashes immediately.

**Devices:** 
- Huawei DUB AL00A 
- Huawei DUB AL00
- Honor BKK AL10
- Huawei ARS AL00
- Huawei DUB AL20

**Android version:** Android 8.1.0,level 27

I noticed that they are all **Huawei phones**, and the EMUI version is 8.x,  some of which using the Qualcomm Snapdragon processor.

----

**updated at 2020-1-3**

After investigating, I found the root cause was the `skia` attemping to use `GL_QCOM_TILED_RENDERING` on these phones. Ssee https://skia-review.googlesource.com/c/skia/+/246282/2. But something went wrong as the logcat said.

If you are troubling on this issue, you can try to roll back the skia revision and [use your local compiled flutter engine](https://github.com/flutter/flutter/wiki/Compiling-the-engine), before the flutter team fixed this issue.

## Logs

```

W libEGL : [eglGetProcAddress] QCOM code detected: glEndTilingQCOM ... Returning NULL
W libEGL : [eglGetProcAddress] QCOM code detected: glStartTilingQCOM ... Returning NULL
E flutter : [ERROR:flutter/shell/gpu/gpu_surface_gl.cc(70)] Failed to setup Skia Gr context.
W libEGL : [eglGetProcAddress] QCOM code detected: glEndTilingQCOM ... Returning NULL
W libEGL : [eglGetProcAddress] QCOM code detected: glStartTilingQCOM ... Returning NULL
```

`main` thread stack trace:
```
1 io.flutter.embedding.engine.FlutterJNI.nativeSurfaceCreated(Native Method)
2 io.flutter.embedding.engine.FlutterJNI.onSurfaceCreated(FlutterJNI.java:3)
3 io.flutter.embedding.engine.renderer.FlutterRenderer.startRenderingToSurface(FlutterRenderer.java:16)
4 io.flutter.embedding.android.FlutterSurfaceView.connectSurfaceToRenderer(FlutterSurfaceView.java:3)
5 io.flutter.embedding.android.FlutterSurfaceView.b(FlutterSurfaceView.java:1)
6 io.flutter.embedding.android.FlutterSurfaceView$a.surfaceCreated(FlutterSurfaceView.java:4)
7 android.view.SurfaceView.updateSurface(SurfaceView.java:728)
8 android.view.SurfaceView$2.onPreDraw(SurfaceView.java:151)
9 android.view.ViewTreeObserver.dispatchOnPreDraw(ViewTreeObserver.java:977)
10 android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2573)
11 android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1558)
12 android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7463)
13 android.view.Choreographer$CallbackRecord.run(Choreographer.java:1041)
14 android.view.Choreographer.doCallbacks(Choreographer.java:847)
15 android.view.Choreographer.doFrame(Choreographer.java:774)
16 android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1027)
17 android.os.Handler.handleCallback(Handler.java:809)
18 android.os.Handler.dispatchMessage(Handler.java:102)
19 android.os.Looper.loop(Looper.java:166)
20 android.app.ActivityThread.main(ActivityThread.java:7555)
```

`1.gpu` thread: signal **SIGSEGV(SEGV_MAPERR)**
```
#00 0x001682a8 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
flutter::Rasterizer::Setup(std::__1::unique_ptr<flutter::Surface, std::__1::default_delete<flutter::Surface> >)
/b/s/w/ir/cache/builder/src/out/android_release/../../flutter/shell/common/rasterizer.cc:69:17
#01 0x0016ec67 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
flutter::Shell::OnPlatformViewCreated(std::__1::unique_ptr<flutter::Surface, std::__1::default_delete<flutter::Surface> >)::$_8::operator()()
/b/s/w/ir/cache/builder/src/out/android_release/../../flutter/shell/common/shell.cc:566:23
auto fml::internal::CopyableLambda<flutter::Shell::OnPlatformViewCreated(std::__1::unique_ptr<flutter::Surface, std::__1::default_delete<flutter::Surface> >)::$_8>::operator()<>() const
/b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/make_copyable.h:24:0
#02 0x00146755 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
std::__1::function<void ()>::operator()() const
/b/s/w/ir/cache/builder/src/out/android_release/../../third_party/libcxx/include/functional:2419:12
fml::MessageLoopImpl::FlushTasks(fml::FlushType)
/b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/message_loop_impl.cc:123:0
fml::MessageLoopImpl::RunExpiredTasksNow()
/b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/message_loop_impl.cc:133:0
#03 0x00148489 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
fml::MessageLoopAndroid::OnEventFired()
/b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:92:5
fml::MessageLoopAndroid::MessageLoopAndroid()::$_0::operator()(int, int, void*) const
/b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:42:0
fml::MessageLoopAndroid::MessageLoopAndroid()::$_0::__invoke(int, int, void*)
/b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:40:0
#04 0x0000ff09 /system/lib/libutils.so (_ZN7android6Looper9pollInnerEi+580) [armeabi-v8::862798b286f1ba97372696549a8bea80]
#05 0x0000fc2d /system/lib/libutils.so (_ZN7android6Looper8pollOnceEiPiS1_PPv+32) [armeabi-v8::862798b286f1ba97372696549a8bea80]
#06 0x0000c9e1 /system/lib/libandroid.so (ALooper_pollOnce+56) [armeabi-v8::2f44f33bd999188a75f397fa44228d46]
#07 0x00148561 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
fml::MessageLoopAndroid::Run()
/b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:68:18
#08 0x00145adf /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
fml::MessageLoopImpl::DoRun()
/b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/message_loop_impl.cc:92:3
fml::MessageLoop::Run()
/b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/message_loop.cc:49:0
#09 0x001474a1 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0::operator()() const
/b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/thread.cc:34:10
decltype(std::__1::forward<fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(fp)()) std::__1::__invoke<fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0&&)
/b/s/w/ir/cache/builder/src/out/android_release/../../third_party/libcxx/include/type_traits:3530:0
void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>&, std::__1::__tuple_indices<>)
/b/s/w/ir/cache/builder/src/out/android_release/../../third_party/libcxx/include/thread:341:0
void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0> >(void*)
/b/s/w/ir/cache/builder/src/out/android_release/../../third_party/libcxx/include/thread:351:0
#10 0x000473eb /system/lib/libc.so (_ZL15__pthread_startPv+22) [armeabi-v8::c38cb0ea3187a942d434b91b139f1edd]
#11 0x0001ad29 /system/lib/libc.so (__start_thread+32) [armeabi-v8::c38cb0ea3187a942d434b91b139f1edd]
```

```
[✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.14.6 18G2022, locale zh-Hans-CN)
    • Flutter version 1.12.13+hotfix.5 at /Users/wangyongrong/Library/flutter
    • Framework revision 27321ebbad (2 weeks ago), 2019-12-10 18:15:01 -0800
    • Engine revision 2994f7e1e6
    • Dart version 2.7.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/wangyongrong/Documents/Android-SDK
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /Users/wangyongrong/Documents/Android-SDK
    • Java binary at: /Applications/Android Studio 2.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] VS Code (version 1.41.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.7.1

[✓] Connected device (1 available)
    • VOG AL10 • APH7N19329001298 • android-arm64 • Android 9 (API 28)

```

**Comment by yrom** [comment]
Workaround for this issue: **enable software rendenring**

Initialize flutter with args:
```java
FlutterMain.ensureInitializationComplete(context, new String[]{FlutterShellArgs.ARG_ENABLE_DART_PROFILING}
```

Or
```java
Intent flutterActivity = ...;
flutterActivity.put(FlutterShellArgs.ARG_KEY_ENABLE_SOFTWARE_RENDERING, true);
startActivity(flutterActivity)
```

**Comment by nuptdzs** [comment]
加了这个会有其他的影响么？

**Comment by yrom** [comment]
> 加了这个会有其他的影响么？

页面会变卡，不过我发现只有EMUI 8.2.0 的手机上会崩，所有只针对这些手机开启这个选项，@nuptdzs

**Comment by nuptdzs** [comment]
> > 加了这个会有其他的影响么？
> 
> 页面会变卡，不过我发现只有EMUI 8.2.0 的手机上会崩，所有只针对这些手机开启这个选项，@nuptdzs
今天上线突然遇到这个问题。只能先回退版本了，机型判断的话担心遗漏不敢这么处理。

**Comment by yrom** [comment]
@nuptdzs 你回退到哪个版本不会崩?

**Comment by nuptdzs** [comment]
> @nuptdzs 你回退到哪个版本不会崩?

1.10.7，理论上1.12以下都行。

**Comment by chg2015** [comment]
> > @nuptdzs你回退到该版本不会崩溃？
> 
> 1.10.7，理论上1.12以下都行。

我也遇到设个问题，怎么回退版本啊

**Comment by iapicca** [comment]
Hi @yrom @nuptdzs @chg2015 
if you update your flutter version recently
you may want to run
`flutter clean`
`flutter pub cache repair`
let me know if the issue persists
thank you

**Comment by ghostgzt** [comment]
me too!

**Comment by haodadaa** [comment]
> Hi @yrom @nuptdzs @chg2015
> if you update your flutter version recently
> you may want to run
> `flutter clean`
> `flutter pub cache repair`
> let me know if 

> Hi @yrom @nuptdzs @chg2015
> if you update your flutter version recently
> you may want to run
> `flutter clean`
> `flutter pub cache repair`
> let me know if the issue persists
> thank you



> Hi @yrom @nuptdzs @chg2015
> if you update your flutter version recently
> you may want to run
> `flutter clean`
> `flutter pub cache repair`
> let me know if the issue persists
> thank you

I also encountered a problem, but this commond can not resolve.

**Comment by haodadaa** [comment]
> Hi @yrom @nuptdzs @chg2015
> if you update your flutter version recently
> you may want to run
> `flutter clean`
> `flutter pub cache repair`
> let me know if the issue persists
> thank you

I also encountered a problem, but this commond can not resolve.

**Comment by zQ4** [comment]
> Hi @yrom @nuptdzs @chg2015
> if you update your flutter version recently
> you may want to run
> `flutter clean`
> `flutter pub cache repair`
> let me know if the issue persists
> thank you



> Hi @yrom @nuptdzs @chg2015
> if you update your flutter version recently
> you may want to run
> `flutter clean`
> `flutter pub cache repair`
> let me know if the issue persists
> thank you

no,荣耀 8c 不通过

**Comment by haodadaa** [comment]
> Hi @yrom @nuptdzs @chg2015
> if you update your flutter version recently
> you may want to run
> `flutter clean`
> `flutter pub cache repair`
> let me know if the issue persists
> thank you

I also encountered a problem, and this command can not resolve.

**Comment by andisolo** [comment]
> Workaround for this issue: **enable software rendering**
> 
> Initialize flutter with args before starting FlutterActivity:
> 
> ```java
> FlutterMain.ensureInitializationComplete(context, new String[]{FlutterShellArgs.ARG_ENABLE_DART_PROFILING});
> ```
> 
> Or
> 
> ```java
> Intent flutterActivity = ...; // build intent
> flutterActivity.put(FlutterShellArgs.ARG_KEY_ENABLE_SOFTWARE_RENDERING, true);
> startActivity(flutterActivity);
> ```
在app的onCreate里面加了这个还是不起作用

**Comment by yrom** [comment]
@andisolo what's the EMUI version on your phone?

**Comment by andisolo** [comment]
> @andisolo what's the EMUI version on your phone?

设备型号:BKK-AL10
系统版本号:BKK-AL10 8.2.0.193(C00)
安卓sdk版本:27
处理器:arm64-v8a
emui版本:EmotionUI_8.2.0

**Comment by yrom** [comment]
@andisolo sorry,I make a mistake! updated the snippet:
```
FlutterMain.ensureInitializationComplete(context, new String[]{FlutterShellArgs.ARG_KEY_ENABLE_SOFTWARE_RENDERING});
```

**Comment by andisolo** [comment]
> @andisolo sorry,I make a mistake! updated the snippet:
> 
> ```
> FlutterMain.ensureInitializationComplete(context, new String[]{FlutterShellArgs.ARG_KEY_ENABLE_SOFTWARE_RENDERING});
> ```

我按照你所说的修改了,现在启动的错误日志如下:
[ERROR:flutter/runtime/dart_vm_data.cc(18)] VM snapshot invalid and could not be inferred from settings.
[ERROR:flutter/runtime/dart_vm.cc(241)] Could not setup VM data to bootstrap the VM from.
[ERROR:flutter/runtime/dart_vm_lifecycle.cc(84)] Could not create Dart VM instance.

**Comment by yrom** [comment]
The `FlutterShellArgs.ARG_KEY_ENABLE_SOFTWARE_RENDERING` should be `FlutterShellArgs.ARG_ENABLE_SOFTWARE_RENDERING`, so sorry..

**Comment by andisolo** [comment]
好了,非常感谢.

**Comment by andisolo** [comment]
> The `FlutterShellArgs.ARG_KEY_ENABLE_SOFTWARE_RENDERING` should be `FlutterShellArgs.ARG_ENABLE_SOFTWARE_RENDERING`, so sorry..

由遇到了和#47765 一样的错误,你知道怎么解决吗?

**Comment by yxwandroid** [comment]
遇到同样的问题

**Comment by yxwandroid** [comment]
@andisolo   你解决了吗

**Comment by zzy216** [comment]
> Workaround for this issue: **enable software rendering**
> 
> Initialize flutter with args before starting FlutterActivity:
> 
> ```java
> FlutterMain.ensureInitializationComplete(context, new String[]{FlutterShellArgs.ARG_ENABLE_SOFTWARE_RENDERING});
> ```
> 
> Or
> 
> ```java
> Intent flutterActivity = ...; // build intent
> flutterActivity.put(FlutterShellArgs.ARG_KEY_ENABLE_SOFTWARE_RENDERING, true);
> startActivity(flutterActivity);
> ```

这个加在哪个文件呀

**Comment by yxwandroid** [comment]
同问

Workaround for this issue: enable software rendering

Initialize flutter with args before starting FlutterActivity:

FlutterMain.ensureInitializationComplete(context, new String[]{FlutterShellArgs.ARG_ENABLE_SOFTWARE_RENDERING});
Or

Intent flutterActivity = ...; // build intent
flutterActivity.put(FlutterShellArgs.ARG_KEY_ENABLE_SOFTWARE_RENDERING, true);
startActivity(flutterActivity);


加在哪

**Comment by andisolo** [comment]
加在app的onCreate生命周期,但是如果有用到原生的view还是会crash详见#47765

**Comment by zzy216** [comment]
> 加在app的onCreate生命周期,但是如果有用到原生的view还是会crash详见#47765

MainActivity.kt的文件下

class MainActivity: FlutterActivity() {
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    Intent flutterActivity = ...; // build intent
    flutterActivity.put(FlutterShellArgs.ARG_KEY_ENABLE_SOFTWARE_RENDERING, true);
    startActivity(flutterActivity);
    GeneratedPluginRegistrant.registerWith(this)
  }
}
这样吗？

**Comment by yxwandroid** [comment]
@andisolo   你有尝试回退flutter 版本解决这个问题吗

**Comment by andisolo** [comment]
> @andisolo 你有尝试回退flutter 版本解决这个问题吗

回退是没有问题的，但是我不想回退。。

**Comment by andisolo** [comment]
> > 加在app的onCreate生命周期,但是如果有用到原生的view还是会crash详见#47765
> 
> MainActivity.kt的文件下
> 
> class MainActivity: FlutterActivity() {
> override fun onCreate(savedInstanceState: Bundle?) {
> super.onCreate(savedInstanceState)
> Intent flutterActivity = ...; // build intent
> flutterActivity.put(FlutterShellArgs.ARG_KEY_ENABLE_SOFTWARE_RENDERING, true);
> startActivity(flutterActivity);
> GeneratedPluginRegistrant.registerWith(this)
> }
> }
> 这样吗？

不是的,自定义app继承FlutterApplication然后在onCreate添加就行了,一定要在父类的onCreate之后

**Comment by ghostgzt** [comment]
> > 加了这个会有其他的影响么？
> > Will this have other effects?

治标不治本，没用的，加了变卡，还不如回退版本

**Comment by hahafather007** [comment]
我滴个法克，我找了半天问题

**Comment by yrom** [comment]
After investigating, I found the root cause was the `skia` attemping to use `GL_QCOM_TILED_RENDERING` on these phones. See <https://skia-review.googlesource.com/c/skia/+/246282/2>. But something went wrong as the logcat said.

```
W libEGL : [eglGetProcAddress] QCOM code detected: glEndTilingQCOM ... Returning NULL
W libEGL : [eglGetProcAddress] QCOM code detected: glStartTilingQCOM ... Returning NULL
```

If you are troubling on this issue, you can try to roll back the skia revision and [use your local compiled flutter engine](https://github.com/flutter/flutter/wiki/Compiling-the-engine), before the flutter team fixed this issue.

The worked skia revision which I have a try is [768f900a88cbfc5237e8edcc16aadde32536ed3e](https://skia.googlesource.com/skia.git/+/768f900a88cbfc5237e8edcc16aadde32536ed3e) 

My `.gclient` file for compiling engine:

```.gclient
solutions = [
  {
    "managed": False,
    "name": "src/flutter",
    "url": "https://github.com/flutter/engine.git@2994f7e1e682039464cb25e31a78b86a3c59b695",
    "custom_deps": {
      'src/third_party/skia': "https://skia.googlesource.com/skia.git@768f900a88cbfc5237e8edcc16aadde32536ed3e",
      'src/third_party/wuffs': 'https://skia.googlesource.com/external/github.com/google/wuffs.git@65e7b2b6c98a4d35e26bc2fc437e2e00f1393dc2',
    },
    "deps_file": "DEPS",
    "safesync_url": "",
  },
]
```

May this comment can help you.

**Comment by mrdong916** [comment]
我同样遇到了这个问题，我回退了版本到v1.9.1+hotfix.6
I also encountered this problem, I rolled back the version to v1.9.1 + hotfix.6

**Comment by hahafather007** [comment]
v1.9.1+hotfix.6 ios无法正常执行flutter run

**Comment by mrdong916** [comment]
> After investigating, I found the root cause was the `skia` attemping to use `GL_QCOM_TILED_RENDERING` on these phones. See https://skia-review.googlesource.com/c/skia/+/246282/2. But something went wrong as the logcat said.
> 
> ```
> W libEGL : [eglGetProcAddress] QCOM code detected: glEndTilingQCOM ... Returning NULL
> W libEGL : [eglGetProcAddress] QCOM code detected: glStartTilingQCOM ... Returning NULL
> ```
> 
> If you are troubling on this issue, you can try to roll back the skia revision and [use your local compiled flutter engine](https://github.com/flutter/flutter/wiki/Compiling-the-engine), before the flutter team fixed this issue.
> 
> The worked skia revision which I have a try is [768f900a88cbfc5237e8edcc16aadde32536ed3e](https://skia.googlesource.com/skia.git/+/768f900a88cbfc5237e8edcc16aadde32536ed3e)
> 
> My `.gclient` file for compiling engine:
> 
> ```
> solutions = [
>   {
>     "managed": False,
>     "name": "src/flutter",
>     "url": "https://github.com/flutter/engine.git@2994f7e1e682039464cb25e31a78b86a3c59b695",
>     "custom_deps": {
>       'src/third_party/skia': "https://skia.googlesource.com/skia.git@768f900a88cbfc5237e8edcc16aadde32536ed3e",
>       'src/third_party/wuffs': 'https://skia.googlesource.com/external/github.com/google/wuffs.git@65e7b2b6c98a4d35e26bc2fc437e2e00f1393dc2',
>     },
>     "deps_file": "DEPS",
>     "safesync_url": "",
>   },
> ]
> ```
> 
> May this comment can help you.

怎么编译和使用这个skia呢？
How to compile and use this skia？

**Comment by zoeyfan** [comment]
@yrom can you consistently repro the issue now? Would you be able to provide a sample project with minimum repro steps?

**Comment by yrom** [comment]
@zoeyfan You can reproduce this issue on the Huawei phone which using Qualcomm processor as mentioned. No explicit steps required, crash on Flutter app startup when built by latest stable Flutter 1.12.13.

**Comment by zoeyfan** [comment]
@cbracken not sure if this is your team's domain. If so, please comment if the description above is sufficient for investigation. If not, please feel free to redirect.

**Comment by rockingdice** [comment]
Please fix this issue asap. I cannot use the latest version of the stable channel anymore. I have to use 1.10.x until the fix arrives.

**Comment by rockingdice** [comment]
Any updates on this? I think a quick fix is to revert the commits which @yrom mentioned.

**Comment by xclud** [comment]
https://github.com/flutter/flutter/issues/18601#issuecomment-558276407

**Comment by zmtzawqlp** [comment]
@yrom could you provide Flutter.framework which works? it is complicated to build the flutter engine and network is a problem when i try to gclient sync. thanks

**Comment by rockingdice** [comment]
@zmtzawqlp I think it's easier to switch back to the older version. I checked the downloaded folder, it's about 16gb..

**Comment by xclud** [comment]
@zmtzawqlp v1.10.4 is the latest working version on Huawei devices. It's a Dev build in SDK Archive.

**Comment by zmtzawqlp** [comment]
@xclud @rockingdice  thanks, but I'm worried about whether there is any problem at old version.

**Comment by yrom** [comment]
@zmtzawqlp  [engine-2994f7e1e682039464cb25e31a78b86a3c59b695.zip](https://github.com/flutter/flutter/files/4038253/engine-2994f7e1e682039464cb25e31a78b86a3c59b695.zip), local repo for android and  only armv7 included, no Flutter.framework.

**Comment by zmtzawqlp** [comment]
@yrom thank you very much ,it's very kind of you.

**Comment by zmtzawqlp** [comment]
@yrom did you do something change for flutter_embedding_release? I just need to replace libflutter.so, right? or also need to use flutter_embedding_release jar in your zip?

**Comment by zmtzawqlp** [comment]
@yrom I am sorry to disturb you again，flutter is still crash in Honor BKK AL10. i notice that size of libflutter.so is much smaller than official(armv7), please take a look when you have a time, thanks.

**Comment by birtanyildiz** [comment]
unfortunately I downgrade flutter to 1.7.8+hotfix.4 for solving the problem.

**Comment by xclud** [comment]
this is the commit which causes the the crash:

https://skia.googlesource.com/skia.git/+/b50cc81a3c211774d0d3c7bf55e8dfbc9061f7cd%5E%21/

I reverted it and it works on my Huawei Honor 8C phone.

**Comment by k00na** [comment]
@birtanyildiz  

Did you try downgrading to higher versions before going all the way to 1.7.8+hotfix.4? Did you maybe try 1.10.x?

**Comment by xclud** [comment]
@k00na 1.10.4 is the most recert version which works. Please see the link of git commit i posted above.

**Comment by birtanyildiz** [comment]
> @k00na 1.10.4 is the most recert version which works. Please see the link of git commit i posted above.

isnt there samsung j4 j6 black screen problem in that version ?

**Comment by xclud** [comment]
> > @k00na 1.10.4 is the most recert version which works. Please see the link of git commit i posted above.
> 
> isnt there samsung j4 j6 black screen problem in that version ?

I dont have any of these models. But it works with galaxy models.

**Comment by yrom** [comment]
> @yrom I am sorry to disturb you again，flutter is still crash in Honor BKK AL10. i notice that size of libflutter.so is much smaller than official(armv7), please take a look when you have a time, thanks.

@zmtzawqlp You mean after overriding `libflutter.so`, your Flutter app crashed on the same issue? Is there only armv7 in your apk? Make sure `libflutter.so` in your apk is overwritten as expected. My app works fine on the same device as you mentioned.

**Comment by zmtzawqlp** [comment]
@yrom thanks, it's my mistake, it works at my side now.

**Comment by GaryQian** [comment]
cc @jason-simmons

**Comment by GaryQian** [comment]
This should be the same as https://github.com/flutter/flutter/issues/47164. I will close the other issue as there seems to be a bit more technical information on this thread. Please continue discussion here and reference the insight found in the other issue!

**Comment by GaryQian** [comment]
We have spoken with Skia, and they will revert the offending patch mentioned above for now to resolve this issue. Will verify on the test devices once they arrive. Thanks everyone for all the research as it made finding the problem much faster!

**Comment by Kuodster** [comment]
> The underlying issue is probably a bad driver or chip from Huawei with an improperly implemented or missing feature, which is difficult to fix on shipped devices. The revert will avoid this issue for the time being.
> 
> I will update this issue as the revert lands/makes its way through the various rollers.

Hello. Won't it make the crash appear again for Samsung's J4 and J6 models?

**Comment by xclud** [comment]
> We have spoken with Skia, and they will revert the offending patch mentioned above for now to resolve this issue. Will verify on the test devices once they arrive. Thanks everyone for all the research as it made finding the problem much faster!

I would appreciate if you let all of us know when the offending patch is reverted (and supposedly roll into flutter).

In reply to another comment of you, if is not only Huawei devices with driver bugs. Investigating Skia source code i found there is a similar bug with Nexus devices as well and the workaround is hard coded in source, something like `if(is_device_nexus()) { ... }`. I am suggesting the similar thing for Huawei devices too.

**Comment by hermanmaritz** [comment]
> @k00na 1.10.4 is the most recert version which works. Please see the link of git commit i posted above.

To save others some time. 1.10.4 is not a good version to roll back to, has a iOS gesture detection bug https://github.com/flutter/flutter/issues/40863

**Comment by xclud** [comment]
@hermanmaritz You can have multiple versions. Currently i build android app with v1.10.4, iOS app with 1.12 hotfix6 and web app with master.

**Comment by k00na** [comment]
@hermanmaritz 
I'm confirming the gesture detection bugs on iOS with 1.10.4. 

@hermanmaritz 
I did exactly the same and it works fine for me as well.

**Comment by dnfield** [comment]
Skia is going to start with reverting the patch. It should roll into Flutter sometime this week. We should see if we can hotfix the skia change into stable, but it's not clear what that will involve just yet (/cc @tvolkert).

As Gary said, the basic problem is that the driver is advertising features it doesn't actually support, and when Skia goes to use those features, it results in a crash. The Skia team is going to figure out how to handle that better once some test devices are available here - for now, the plan is to just revert.

**Comment by tvolkert** [comment]
> it's not clear what that will involve just yet

We'd need the Skia team to cherrypick the revert into a branch (off of commit `c3ff97a98b74554854b5822797c6b7b431f4d8d2`), then we'd cherrypick that new Skia commit into the engine on a hotfixes branch.

@bsalomon is this something Skia team would be willing to do?

**Comment by geeksammao** [comment]
@GaryQian Is there any quick fix for this issue except rolling back flutter version(we use some flutter 1.12 feature and don't want to rewrite the code)? Since we need to upload our apk this week, in case the hotfix version can't drop in time, I wonder if there is any patch we can apply by customizing local flutter engine or any commit we can revert manually.

**Comment by rockingdice** [comment]
@geeksammao I managed to deal with it with the following method:

1. Follow the guide of @yrom provided, you'll get all local engines for different architectures. You will need the release version if you want to make APKs.
2. Make an 'original.apk' using the original engines. (include all of the abi), unpack it with apktool.
3. Make a 'modified.apk' using one of your local engines. (--local-engine doesn't support for multiple targets). You have to limit the abi to the only one you use. 
4. Extract those modified .so (/lib/arm64-v8a/* as an example) from the APK bundle, and replace the old one in the original.apk. 
5. Repeat 3 and 4 steps for other abi.
6. Pack your original.apk with apktool and resign it with apksigner.

I tested it myself and it's working.

**Comment by geeksammao** [comment]
> @geeksammao I managed to deal with it with the following method:
> 
> 1. Follow the guide of @yrom provided, you'll get all local engines for different architectures. You will need the release version if you want to make APKs.
> 2. Make an 'original.apk' using the original engines. (include all of the abi), unpack it with apktool.
> 3. Make a 'modified.apk' using one of your local engines. (--local-engine doesn't support for multiple targets). You have to limit the abi to the only one you use.
> 4. Extract those modified .so (/lib/arm64-v8a/* as an example) from the APK bundle, and replace the old one in the original.apk.
> 5. Repeat 3 and 4 steps for other abi.
> 6. Pack your original.apk with apktool and resign it with apksigner.
> 
> I tested it myself and it's working.

You mean rolling back Skia with local engine compiled? I have done it but I can't confirm it doesn't have any other issues, so for now I pack 2 libflutter.so(one original, one with Skia rolled back) in one apk and only use the customized libflutter.so for those Huawei users. I think this is a dirty fix, so I want  to find a better one.

**Comment by geeksammao** [comment]
> @geeksammao I managed to deal with it with the following method:
> 
> 1. Follow the guide of @yrom provided, you'll get all local engines for different architectures. You will need the release version if you want to make APKs.
> 2. Make an 'original.apk' using the original engines. (include all of the abi), unpack it with apktool.
> 3. Make a 'modified.apk' using one of your local engines. (--local-engine doesn't support for multiple targets). You have to limit the abi to the only one you use.
> 4. Extract those modified .so (/lib/arm64-v8a/* as an example) from the APK bundle, and replace the old one in the original.apk.
> 5. Repeat 3 and 4 steps for other abi.
> 6. Pack your original.apk with apktool and resign it with apksigner.
> 
> I tested it myself and it's working.

You mean compiling local engine with Skia rolled back? I have done it but I can't confirm it doesn't have any other issues, so for now I pack 2 libflutter.so(one original, one with Skia rolled back) in one apk and only use the customized libflutter.so for those Huawei users. I think this is a dirty fix, so I want  to find a better one.

**Comment by geeksammao** [comment]
> @geeksammao I managed to deal with it with the following method:
> 
> 1. Follow the guide of @yrom provided, you'll get all local engines for different architectures. You will need the release version if you want to make APKs.
> 2. Make an 'original.apk' using the original engines. (include all of the abi), unpack it with apktool.
> 3. Make a 'modified.apk' using one of your local engines. (--local-engine doesn't support for multiple targets). You have to limit the abi to the only one you use.
> 4. Extract those modified .so (/lib/arm64-v8a/* as an example) from the APK bundle, and replace the old one in the original.apk.
> 5. Repeat 3 and 4 steps for other abi.
> 6. Pack your original.apk with apktool and resign it with apksigner.
> 
> I tested it myself and it's working.

You mean rolling back Skia with local engine compiled? I have done it but I can't confirm it doesn't have any other issues, so for now I pack 2 libflutter.so(one original, one with Skia rolled back) in one apk and only use the customized libflutter.so for those Huawei users. I think this is a dirty fix, so I want  to find a better one.

**Comment by nuptdzs** [comment]
> 
> 经历过这个BUG，公司内部给出了一个结论，为了保证线上用户使用不出现意外的崩溃问题：
> 1、无重大更新或者问题性修复时，不进行flutter 版本的更新；
> 2、内部需要更新flutter版本时，需要提前进行版本兼容性测试，覆盖APPtop机型，测试通过情况下方可使用flutter新版本进行开发。避免使用了新版开发测试出问题需要回退版本的问题。
After experiencing this bug, the company came to a conclusion that in order to ensure that online users do not have unexpected crash problems:
1. When there is no major update or problem repair, the version of flitter will not be updated;
2. When the internal version of flutter needs to be updated, the version compatibility test should be carried out in advance to cover the apptop model. If the test passes, the new version of flutter can be used for development. To avoid using the new version of development testing problems need to backoff the version of the problem.

**Comment by xclud** [comment]
It seems the bug is fixed in Skia now. I cant wait for the next flutter engine roll.

Here are the details:
https://skia.googlesource.com/skia.git/+/0a1923e50752073fd31ad667935c9e37681917e8

**Comment by xclud** [comment]
It seems the bug is fixed in Skia now. I cant wait for the next flutter engine roll.

Here are the details:
https://skia.googlesource.com/skia.git/+/0a1923e50752073fd31ad667935c9e37681917e8%5E%21/

**Comment by xclud** [comment]
Skia rolled in https://github.com/flutter/flutter/pull/48733

I am happy to confirm v1.13.9-pre.63 is working on my Huawei mobile.

Thank you all for the great support.

**Comment by xclud** [comment]
Skia rolled in https://github.com/flutter/flutter/pull/48733

I am happy to confirm this PR is working on my Huawei mobile.

Thank you all for the great support.

**Comment by GaryQian** [comment]
Thanks for confirming! Appreciate the all the sleuthing everyone did, made the revert/fix a breeze. We have test devices coming soon, and will try to reland the reverted changes with confirmation that the bug is avoided.

I will close this for now, please feel free to continue providing feedback if relevant!

**Comment by AlexVincent525** [comment]
Is there going to be a hotfix or something?

**Comment by rockingdice** [comment]
@geeksammao  You should use the master channel instead now.

**Comment by tvolkert** [comment]
@AlexVincent525 I'm investigating putting together a hotfix (I'll need help from the Skia team creating the cherry-pick on a branch in their repo).

**Comment by mahmood-heja** [comment]
Same error here on flutter v1.13.8, on device Huawei Y7 2019 (android 8.1) 

```
> backtrace:
>   #00  pc 00000000006ce3cc  /data/app/com.tasheh.tasheheats-sGEYSBMSpOJJyUV2IPgJSQ==/lib/arm64/libflutter.so (offset 0x270000)
>   #01  pc 00000000002b7be4  /data/app/com.tasheh.tasheheats-sGEYSBMSpOJJyUV2IPgJSQ==/lib/arm64/libflutter.so (offset 0x270000)
>   #02  pc 00000000002bf744  /data/app/com.tasheh.tasheheats-sGEYSBMSpOJJyUV2IPgJSQ==/lib/arm64/libflutter.so (offset 0x270000)
>   #03  pc 0000000000290000  /data/app/com.tasheh.tasheheats-sGEYSBMSpOJJyUV2IPgJSQ==/lib/arm64/libflutter.so (offset 0x270000)
>   #04  pc 00000000002926ec  /data/app/com.tasheh.tasheheats-sGEYSBMSpOJJyUV2IPgJSQ==/lib/arm64/libflutter.so (offset 0x270000)
>   #05  pc 0000000000016238  /system/lib64/libutils.so (android::Looper::pollInner(int)+816)
>   #06  pc 0000000000015e78  /system/lib64/libutils.so (android::Looper::pollOnce(int, int*, int*, void**)+108)
>   #07  pc 0000000000011854  /system/lib64/libandroid.so (ALooper_pollOnce+92)
>   #08  pc 0000000000292808  /data/app/com.tasheh.tasheheats-sGEYSBMSpOJJyUV2IPgJSQ==/lib/arm64/libflutter.so (offset 0x270000)
>   #09  pc 000000000028f1d8  /data/app/com.tasheh.tasheheats-sGEYSBMSpOJJyUV2IPgJSQ==/lib/arm64/libflutter.so (offset 0x270000)
>   #10  pc 0000000000291290  /data/app/com.tasheh.tasheheats-sGEYSBMSpOJJyUV2IPgJSQ==/lib/arm64/libflutter.so (offset 0x270000)
>   #11  pc 0000000000077e9c  /system/lib64/libc.so (__pthread_start(void*)+36)
>   #12  pc 000000000001f6c0  /system/lib64/libc.so (__start_thread+68)
> 
```

**Comment by zmtzawqlp** [comment]
@tvolkert any update for the hotfix? we are all waiting for it. thanks.

**Comment by xclud** [comment]
@mahmood-heja run `flutter pub upgrade` in the project folder.

**Comment by xclud** [comment]
@mahmood-heja v1.13.9-pre63 is the version which fixed the bug in skia.

After upgrading to last version in master channel, run `flutter pub upgrade`.

I hope this does the trick.

**Comment by mahmood-heja** [comment]
> @mahmood-heja run `flutter pub upgrade` in the project folder.

thx bro, i will do that

**Comment by JaonFanwt** [comment]
iOS crash too, but not often.
(release, iPhone 11, iOS:13.3 (17C54), arm64-v8a)

v1.12.13+hotfix.5
Engine revision 2994f7e1e6

```
#5 Thread
SIGSEGV
SEGV_ACCERR
0 Flutter	0x0000000103378644 0x00000001032f4000 + 542276
1 Flutter	0x0000000103378640 0x00000001032f4000 + 542272
2 Flutter	0x000000010332f530 0x00000001032f4000 + 242992
3 Flutter	0x0000000103331b9c 0x00000001032f4000 + 252828
4 CoreFoundation	0x0000000183713aa8 0x0000000183625000 + 977576
5 CoreFoundation	0x000000018371376c 0x0000000183625000 + 976748
6 CoreFoundation	0x0000000183713010 0x0000000183625000 + 974864
7 CoreFoundation	0x0000000183710b60 0x0000000183625000 + 965472
8 CoreFoundation	0x0000000183630da8 CFRunLoopRunSpecific + 552
9 Flutter	0x0000000103331a78 0x00000001032f4000 + 252536
10 Flutter	0x00000001033311e4 0x00000001032f4000 + 250340
11 libsystem_pthread.dylib	0x0000000183391220 0x000000018338f000 + 8736
12 libsystem_pthread.dylib	0x0000000183391110 0x000000018338f000 + 8464
```
-------- atos --------
```
0 auto fml::internal::CopyableLambda<flutter::Shell::OnPlatformViewCreated(std::__1::unique_ptr<flutter::Surface, std::__1::default_delete<flutter::Surface> >)::$_8>::operator()<>() const (in Flutter) (make_copyable.h:24)
1 auto fml::internal::CopyableLambda<flutter::Shell::OnPlatformViewCreated(std::__1::unique_ptr<flutter::Surface, std::__1::default_delete<flutter::Surface> >)::$_8>::operator()<>() const (in Flutter) (make_copyable.h:24)
2 fml::MessageLoopImpl::FlushTasks(fml::FlushType) (in Flutter) (message_loop_impl.cc:125)
3 fml::MessageLoopDarwin::OnTimerFire(__CFRunLoopTimer*, fml::MessageLoopDarwin*) (in Flutter) (message_loop_darwin.mm:76)
9 fml::MessageLoopDarwin::Run() (in Flutter) (message_loop_darwin.mm:47)
10 void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0> >(void*) (in Flutter) (thread:352)
```

**Comment by tvolkert** [comment]
Hotfix process has begun in https://github.com/flutter/engine/pull/15700

**Comment by V1rus999** [comment]
Has anyone else noticed significant performance degradation when changing to the master branch?

The master branch did fix the crash issue on my Huawei Y7, but now our app is basically unusable. Plus were seeing significant performance issues on non Huawei devices and emulators.

**Comment by dnfield** [comment]
@V1rus999 please file a new bug for any performance regresisons you're seeing.  Include as many relevant details as you can like:

- Last known revision where the performance was better
- Known revision where preformance degrades
- A minimal example we can run to observe the performance degredation
- Any hardware information for what you're testing on (but note that we tend to not prioritize performance on emulators/simulators).

**Comment by tvolkert** [comment]
Cherry-picks into framework repo in preparation for the hotfix is at https://github.com/flutter/flutter/pull/49437

**Comment by da7om85** [comment]
hello everyone,
new stable version has released, v1.12.13+hotfix.7-stable.
Can anyone check if the problem with Huawei is fixed in this version or not

**Comment by V1rus999** [comment]
> 
> 
> hello everyone,
> new stable version has released, v1.12.13+hotfix.7-stable.
> Can anyone check if the problem with Huawei is fixed in this version or not

**I can confirm that this runs on my Huawei Y7 Prime, BUT the performance has degraded significantly.** 

* Our app is almost unusable. There are clear animation stuttering and the app takes really long to start up.
* **Even the flutter demo counter app has clear animations issues and takes about 5 seconds to open**

Flutter doctor:

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, v1.12.13+hotfix.7, on Microsoft Windows [Version 10.0.18362.592], locale en-US)
 
[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[√] Android Studio (version 3.5)
[!] IntelliJ IDEA Community Edition (version 2019.1)
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
[√] VS Code (version 1.41.1)
[√] Connected device (1 available)

! Doctor found issues in 1 category.
```

**Comment by tvolkert** [comment]
@V1rus999 can you file a separate issue about the performance regression on Huawei Y7 and tag @liyuqian?

**Comment by tvolkert** [comment]
The performance issues on those phones may also be improved by https://skia-review.googlesource.com/c/skia/+/264695, which will be included in the next release.

**Comment by xclud** [comment]
> hello everyone,
> new stable version has released, v1.12.13+hotfix.7-stable.
> Can anyone check if the problem with Huawei is fixed in this version or not

Its significantly slow on my Huawei Honor 8C.

**Comment by liyuqian** [comment]
@godofredoc @GaryQian : we might need to add the 2 Huawei device to our device lab to continuously monitor this performance regression.

**Comment by iapicca** [comment]
Could everyone who still has this problem please file a new issue with the exact descriptions what happens, logs and the output of 'flutter doctor -v'  please.
All system setups can be slightly different so it's always better to open new issues and reference related issues.

**Comment by rvhamed** [comment]
I upgraded the flutter and after that run flutter pub upgrade, but I still have this issue.

`
E/flutter ( 3316): [ERROR:flutter/shell/gpu/gpu_surface_gl.cc(70)] Failed to setup Skia Gr context.
`

 flutter doctor -v
[✓] Flutter (Channel stable, v1.12.13+hotfix.7, on Mac OS X 10.14.6 18G103, locale en-US)
    • Flutter version 1.12.13+hotfix.7 at /Users/home/Develop/SDK_Flutter
    • Framework revision 9f5ff2306b (10 days ago), 2020-01-26 22:38:26 -0800
    • Engine revision a67792536c
    • Dart version 2.7.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/home/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.1, Build version 11A1027
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 42.1.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] Connected device (1 available)
    • HTC One X, 4 2 2, API 17, 720x1280 • 192.168.56.102:5555 • android-x86 • Android 4.2.2 (API 17)

• No issues found!

**Comment by amrosama2020** [comment]
> The performance issues on those phones may also be improved by https://skia.googlesource.com/skia/+/1321a3d, which will be included in the next release.

Any idea when the next release would be? this info would be helpful to decide on what to do.

**Comment by liyuqian** [comment]
@amrosama2020 : Flutter has a new release version roughly every quarter (3 months). Our last one is around December so the next one should be around March.

**Comment by zsunkun** [comment]
I still have this problem with version 1.12.13+hotfix8,Can not be reproduced locally.
As @singun-lxd mentioned,[#51447](https://github.com/flutter/flutter/issues/51447)，I think they have the same problem.

**Comment by Hixie** [comment]
cc @jason-simmons Any chance you can take a look at this one? Thanks!

**Comment by zsunkun** [comment]
> @zsunkun, @WATER1350, @linxuebin1990,@Qxyat. Would you please provide a particular device model and its OS version so our team can look into this issue?

@jason-simmons @zoeyfan  
device model:
![image](https://user-images.githubusercontent.com/5997059/79936495-3c492000-848a-11ea-98fe-05835e4ce574.png)
os version:
![image](https://user-images.githubusercontent.com/5997059/79936561-61d62980-848a-11ea-9863-6324c91b143a.png)

**Comment by GaryQian** [comment]
Although the crash error is the same, it looks to me like this may have a different cause than the original bug filed here since this seems to be more OS specific as well as device specific.

I'm going to merge the continuation of this issue with #51447 and close this one.

@zsunkun I will reproduce your useful graphs in the other issue. Thanks!

**Comment by lionoggo** [comment]
> ## Steps to Reproduce
> This crash was happened on users' Android phone. I don't know clear steps to reproduce.
> 
> User starts the `FlutterActivity` then the APP crashes immediately.
> 
> **Devices:**
> 
> * Huawei DUB AL00A
> * Huawei DUB AL00
> * Honor BKK AL10
> * Huawei ARS AL00
> * Huawei DUB AL20
> 
> **Android version:** Android 8.1.0,level 27
> 
> I noticed that they are all **Huawei phones**, and the EMUI version is 8.x, some of which using the Qualcomm Snapdragon processor.
> 
> **updated at 2020-1-3**
> 
> After investigating, I found the root cause was the `skia` attemping to use `GL_QCOM_TILED_RENDERING` on these phones. Ssee https://skia-review.googlesource.com/c/skia/+/246282/2. But something went wrong as the logcat said.
> 
> If you are troubling on this issue, you can try to roll back the skia revision and [use your local compiled flutter engine](https://github.com/flutter/flutter/wiki/Compiling-the-engine), before the flutter team fixed this issue.
> 
> ## Logs
> ```
> 
> W libEGL : [eglGetProcAddress] QCOM code detected: glEndTilingQCOM ... Returning NULL
> W libEGL : [eglGetProcAddress] QCOM code detected: glStartTilingQCOM ... Returning NULL
> E flutter : [ERROR:flutter/shell/gpu/gpu_surface_gl.cc(70)] Failed to setup Skia Gr context.
> W libEGL : [eglGetProcAddress] QCOM code detected: glEndTilingQCOM ... Returning NULL
> W libEGL : [eglGetProcAddress] QCOM code detected: glStartTilingQCOM ... Returning NULL
> ```
> 
> `main` thread stack trace:
> 
> ```
> 1 io.flutter.embedding.engine.FlutterJNI.nativeSurfaceCreated(Native Method)
> 2 io.flutter.embedding.engine.FlutterJNI.onSurfaceCreated(FlutterJNI.java:3)
> 3 io.flutter.embedding.engine.renderer.FlutterRenderer.startRenderingToSurface(FlutterRenderer.java:16)
> 4 io.flutter.embedding.android.FlutterSurfaceView.connectSurfaceToRenderer(FlutterSurfaceView.java:3)
> 5 io.flutter.embedding.android.FlutterSurfaceView.b(FlutterSurfaceView.java:1)
> 6 io.flutter.embedding.android.FlutterSurfaceView$a.surfaceCreated(FlutterSurfaceView.java:4)
> 7 android.view.SurfaceView.updateSurface(SurfaceView.java:728)
> 8 android.view.SurfaceView$2.onPreDraw(SurfaceView.java:151)
> 9 android.view.ViewTreeObserver.dispatchOnPreDraw(ViewTreeObserver.java:977)
> 10 android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2573)
> 11 android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1558)
> 12 android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7463)
> 13 android.view.Choreographer$CallbackRecord.run(Choreographer.java:1041)
> 14 android.view.Choreographer.doCallbacks(Choreographer.java:847)
> 15 android.view.Choreographer.doFrame(Choreographer.java:774)
> 16 android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1027)
> 17 android.os.Handler.handleCallback(Handler.java:809)
> 18 android.os.Handler.dispatchMessage(Handler.java:102)
> 19 android.os.Looper.loop(Looper.java:166)
> 20 android.app.ActivityThread.main(ActivityThread.java:7555)
> ```
> 
> `1.gpu` thread: signal **SIGSEGV(SEGV_MAPERR)**
> 
> ```
> #00 0x001682a8 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> flutter::Rasterizer::Setup(std::__1::unique_ptr<flutter::Surface, std::__1::default_delete<flutter::Surface> >)
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/shell/common/rasterizer.cc:69:17
> #01 0x0016ec67 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> flutter::Shell::OnPlatformViewCreated(std::__1::unique_ptr<flutter::Surface, std::__1::default_delete<flutter::Surface> >)::$_8::operator()()
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/shell/common/shell.cc:566:23
> auto fml::internal::CopyableLambda<flutter::Shell::OnPlatformViewCreated(std::__1::unique_ptr<flutter::Surface, std::__1::default_delete<flutter::Surface> >)::$_8>::operator()<>() const
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/make_copyable.h:24:0
> #02 0x00146755 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> std::__1::function<void ()>::operator()() const
> /b/s/w/ir/cache/builder/src/out/android_release/../../third_party/libcxx/include/functional:2419:12
> fml::MessageLoopImpl::FlushTasks(fml::FlushType)
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/message_loop_impl.cc:123:0
> fml::MessageLoopImpl::RunExpiredTasksNow()
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/message_loop_impl.cc:133:0
> #03 0x00148489 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> fml::MessageLoopAndroid::OnEventFired()
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:92:5
> fml::MessageLoopAndroid::MessageLoopAndroid()::$_0::operator()(int, int, void*) const
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:42:0
> fml::MessageLoopAndroid::MessageLoopAndroid()::$_0::__invoke(int, int, void*)
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:40:0
> #04 0x0000ff09 /system/lib/libutils.so (_ZN7android6Looper9pollInnerEi+580) [armeabi-v8::862798b286f1ba97372696549a8bea80]
> #05 0x0000fc2d /system/lib/libutils.so (_ZN7android6Looper8pollOnceEiPiS1_PPv+32) [armeabi-v8::862798b286f1ba97372696549a8bea80]
> #06 0x0000c9e1 /system/lib/libandroid.so (ALooper_pollOnce+56) [armeabi-v8::2f44f33bd999188a75f397fa44228d46]
> #07 0x00148561 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> fml::MessageLoopAndroid::Run()
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:68:18
> #08 0x00145adf /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> fml::MessageLoopImpl::DoRun()
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/message_loop_impl.cc:92:3
> fml::MessageLoop::Run()
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/message_loop.cc:49:0
> #09 0x001474a1 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0::operator()() const
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/thread.cc:34:10
> decltype(std::__1::forward<fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(fp)()) std::__1::__invoke<fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0&&)
> /b/s/w/ir/cache/builder/src/out/android_release/../../third_party/libcxx/include/type_traits:3530:0
> void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>&, std::__1::__tuple_indices<>)
> /b/s/w/ir/cache/builder/src/out/android_release/../../third_party/libcxx/include/thread:341:0
> void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0> >(void*)
> /b/s/w/ir/cache/builder/src/out/android_release/../../third_party/libcxx/include/thread:351:0
> #10 0x000473eb /system/lib/libc.so (_ZL15__pthread_startPv+22) [armeabi-v8::c38cb0ea3187a942d434b91b139f1edd]
> #11 0x0001ad29 /system/lib/libc.so (__start_thread+32) [armeabi-v8::c38cb0ea3187a942d434b91b139f1edd]
> ```
> 
> ```
> [✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.14.6 18G2022, locale zh-Hans-CN)
>     • Flutter version 1.12.13+hotfix.5 at /Users/wangyongrong/Library/flutter
>     • Framework revision 27321ebbad (2 weeks ago), 2019-12-10 18:15:01 -0800
>     • Engine revision 2994f7e1e6
>     • Dart version 2.7.0
> 
>  
> [✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
>     • Android SDK at /Users/wangyongrong/Documents/Android-SDK
>     • Android NDK location not configured (optional; useful for native profiling support)
>     • Platform android-29, build-tools 29.0.2
>     • ANDROID_HOME = /Users/wangyongrong/Documents/Android-SDK
>     • Java binary at: /Applications/Android Studio 2.app/Contents/jre/jdk/Contents/Home/bin/java
>     • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
>     • All Android licenses accepted.
> 
> [✓] VS Code (version 1.41.1)
>     • VS Code at /Applications/Visual Studio Code.app/Contents
>     • Flutter extension version 3.7.1
> 
> [✓] Connected device (1 available)
>     • VOG AL10 • APH7N19329001298 • android-arm64 • Android 9 (API 28)
> ```



> ## Steps to Reproduce
> This crash was happened on users' Android phone. I don't know clear steps to reproduce.
> 
> User starts the `FlutterActivity` then the APP crashes immediately.
> 
> **Devices:**
> 
> * Huawei DUB AL00A
> * Huawei DUB AL00
> * Honor BKK AL10
> * Huawei ARS AL00
> * Huawei DUB AL20
> 
> **Android version:** Android 8.1.0,level 27
> 
> I noticed that they are all **Huawei phones**, and the EMUI version is 8.x, some of which using the Qualcomm Snapdragon processor.
> 
> **updated at 2020-1-3**
> 
> After investigating, I found the root cause was the `skia` attemping to use `GL_QCOM_TILED_RENDERING` on these phones. Ssee https://skia-review.googlesource.com/c/skia/+/246282/2. But something went wrong as the logcat said.
> 
> If you are troubling on this issue, you can try to roll back the skia revision and [use your local compiled flutter engine](https://github.com/flutter/flutter/wiki/Compiling-the-engine), before the flutter team fixed this issue.
> 
> ## Logs
> ```
> 
> W libEGL : [eglGetProcAddress] QCOM code detected: glEndTilingQCOM ... Returning NULL
> W libEGL : [eglGetProcAddress] QCOM code detected: glStartTilingQCOM ... Returning NULL
> E flutter : [ERROR:flutter/shell/gpu/gpu_surface_gl.cc(70)] Failed to setup Skia Gr context.
> W libEGL : [eglGetProcAddress] QCOM code detected: glEndTilingQCOM ... Returning NULL
> W libEGL : [eglGetProcAddress] QCOM code detected: glStartTilingQCOM ... Returning NULL
> ```
> 
> `main` thread stack trace:
> 
> ```
> 1 io.flutter.embedding.engine.FlutterJNI.nativeSurfaceCreated(Native Method)
> 2 io.flutter.embedding.engine.FlutterJNI.onSurfaceCreated(FlutterJNI.java:3)
> 3 io.flutter.embedding.engine.renderer.FlutterRenderer.startRenderingToSurface(FlutterRenderer.java:16)
> 4 io.flutter.embedding.android.FlutterSurfaceView.connectSurfaceToRenderer(FlutterSurfaceView.java:3)
> 5 io.flutter.embedding.android.FlutterSurfaceView.b(FlutterSurfaceView.java:1)
> 6 io.flutter.embedding.android.FlutterSurfaceView$a.surfaceCreated(FlutterSurfaceView.java:4)
> 7 android.view.SurfaceView.updateSurface(SurfaceView.java:728)
> 8 android.view.SurfaceView$2.onPreDraw(SurfaceView.java:151)
> 9 android.view.ViewTreeObserver.dispatchOnPreDraw(ViewTreeObserver.java:977)
> 10 android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2573)
> 11 android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1558)
> 12 android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7463)
> 13 android.view.Choreographer$CallbackRecord.run(Choreographer.java:1041)
> 14 android.view.Choreographer.doCallbacks(Choreographer.java:847)
> 15 android.view.Choreographer.doFrame(Choreographer.java:774)
> 16 android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1027)
> 17 android.os.Handler.handleCallback(Handler.java:809)
> 18 android.os.Handler.dispatchMessage(Handler.java:102)
> 19 android.os.Looper.loop(Looper.java:166)
> 20 android.app.ActivityThread.main(ActivityThread.java:7555)
> ```
> 
> `1.gpu` thread: signal **SIGSEGV(SEGV_MAPERR)**
> 
> ```
> #00 0x001682a8 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> flutter::Rasterizer::Setup(std::__1::unique_ptr<flutter::Surface, std::__1::default_delete<flutter::Surface> >)
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/shell/common/rasterizer.cc:69:17
> #01 0x0016ec67 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> flutter::Shell::OnPlatformViewCreated(std::__1::unique_ptr<flutter::Surface, std::__1::default_delete<flutter::Surface> >)::$_8::operator()()
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/shell/common/shell.cc:566:23
> auto fml::internal::CopyableLambda<flutter::Shell::OnPlatformViewCreated(std::__1::unique_ptr<flutter::Surface, std::__1::default_delete<flutter::Surface> >)::$_8>::operator()<>() const
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/make_copyable.h:24:0
> #02 0x00146755 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> std::__1::function<void ()>::operator()() const
> /b/s/w/ir/cache/builder/src/out/android_release/../../third_party/libcxx/include/functional:2419:12
> fml::MessageLoopImpl::FlushTasks(fml::FlushType)
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/message_loop_impl.cc:123:0
> fml::MessageLoopImpl::RunExpiredTasksNow()
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/message_loop_impl.cc:133:0
> #03 0x00148489 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> fml::MessageLoopAndroid::OnEventFired()
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:92:5
> fml::MessageLoopAndroid::MessageLoopAndroid()::$_0::operator()(int, int, void*) const
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:42:0
> fml::MessageLoopAndroid::MessageLoopAndroid()::$_0::__invoke(int, int, void*)
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:40:0
> #04 0x0000ff09 /system/lib/libutils.so (_ZN7android6Looper9pollInnerEi+580) [armeabi-v8::862798b286f1ba97372696549a8bea80]
> #05 0x0000fc2d /system/lib/libutils.so (_ZN7android6Looper8pollOnceEiPiS1_PPv+32) [armeabi-v8::862798b286f1ba97372696549a8bea80]
> #06 0x0000c9e1 /system/lib/libandroid.so (ALooper_pollOnce+56) [armeabi-v8::2f44f33bd999188a75f397fa44228d46]
> #07 0x00148561 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> fml::MessageLoopAndroid::Run()
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/platform/android/message_loop_android.cc:68:18
> #08 0x00145adf /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> fml::MessageLoopImpl::DoRun()
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/message_loop_impl.cc:92:3
> fml::MessageLoop::Run()
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/message_loop.cc:49:0
> #09 0x001474a1 /data/app/com.bilibili.comic-UlhrVZRG20zydm5ToI41uw==/lib/arm/libflutter.so [armeabi-v7a::8d3ed6bbd864323d41970181b0b00b84]
> fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0::operator()() const
> /b/s/w/ir/cache/builder/src/out/android_release/../../flutter/fml/thread.cc:34:10
> decltype(std::__1::forward<fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(fp)()) std::__1::__invoke<fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0&&)
> /b/s/w/ir/cache/builder/src/out/android_release/../../third_party/libcxx/include/type_traits:3530:0
> void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>&, std::__1::__tuple_indices<>)
> /b/s/w/ir/cache/builder/src/out/android_release/../../third_party/libcxx/include/thread:341:0
> void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0> >(void*)
> /b/s/w/ir/cache/builder/src/out/android_release/../../third_party/libcxx/include/thread:351:0
> #10 0x000473eb /system/lib/libc.so (_ZL15__pthread_startPv+22) [armeabi-v8::c38cb0ea3187a942d434b91b139f1edd]
> #11 0x0001ad29 /system/lib/libc.so (__start_thread+32) [armeabi-v8::c38cb0ea3187a942d434b91b139f1edd]
> ```
> 
> ```
> [✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.14.6 18G2022, locale zh-Hans-CN)
>     • Flutter version 1.12.13+hotfix.5 at /Users/wangyongrong/Library/flutter
>     • Framework revision 27321ebbad (2 weeks ago), 2019-12-10 18:15:01 -0800
>     • Engine revision 2994f7e1e6
>     • Dart version 2.7.0
> 
>  
> [✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
>     • Android SDK at /Users/wangyongrong/Documents/Android-SDK
>     • Android NDK location not configured (optional; useful for native profiling support)
>     • Platform android-29, build-tools 29.0.2
>     • ANDROID_HOME = /Users/wangyongrong/Documents/Android-SDK
>     • Java binary at: /Applications/Android Studio 2.app/Contents/jre/jdk/Contents/Home/bin/java
>     • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
>     • All Android licenses accepted.
> 
> [✓] VS Code (version 1.41.1)
>     • VS Code at /Applications/Visual Studio Code.app/Contents
>     • Flutter extension version 3.7.1
> 
> [✓] Connected device (1 available)
>     • VOG AL10 • APH7N19329001298 • android-arm64 • Android 9 (API 28)
> ```

@yrom  How to specify the local skia when compiling the engine?
